yt/geometry/particle_oct_container.pyx

```diff
@@ -985,47 +994,38 @@ cdef class ParticleBitmap:
         nc, nm = self.bitmasks._find_collisions_refined(self.collisions,verbose)
         return nc, nm
 
-    def calcsize_bitmasks(self):
-        # TODO: All cython
-        cdef bytes serial_BAC
-        cdef np.uint64_t ifile
-        cdef int out = 0
-        out += struct.calcsize('Q')
-        # Bitmaps for each file
-        for ifile in range(self.nfiles):
-            serial_BAC = self.bitmasks._dumps(ifile)
-            out += struct.calcsize('Q')
-            out += len(serial_BAC)
-        # Bitmap for collisions
-        serial_BAC = self.collisions._dumps()
-        out += struct.calcsize('Q')
-        out += len(serial_BAC)
-        return out
-
     def get_bitmasks(self):
         return self.bitmasks
 
     def iseq_bitmask(self, solf):
         return self.bitmasks._iseq(solf.get_bitmasks())
 
     def save_bitmasks(self, fname):
+        import h5py
         cdef bytes serial_BAC
         cdef np.uint64_t ifile
-        f = open(fname,'wb')
-        # Header
-        f.write(struct.pack('Q', _bitmask_version))
-        f.write(struct.pack('q', self.file_hash))
-        f.write(struct.pack('Q', self.nfiles))
-        # Bitmap for each file
-        for ifile in range(self.nfiles):
-            serial_BAC = self.bitmasks._dumps(ifile)
-            f.write(struct.pack('Q', len(serial_BAC)))
-            f.write(serial_BAC)
-        # Collisions
-        serial_BAC = self.collisions._dumps()
-        f.write(struct.pack('Q', len(serial_BAC)))
-        f.write(serial_BAC)
-        f.close()
+        with h5py.File(fname, mode="a") as fp:
+            try:
+                grp = fp[str(self.hash_value)]
+                grp.clear()
+            except KeyError:
+                grp = fp.create_group(str(self.hash_value))
+
+            grp.attrs["bitmask_version"] = _bitmask_version
+            grp.attrs["nfiles"] = self.nfiles
+            # Add some attrs for convenience. They're not read back.
+            grp.attrs["file_hash"] = self.file_hash
+            grp.attrs["left_edge"] = self.left_edge
+            grp.attrs["right_edge"] = self.right_edge
+            grp.attrs["periodicity"] = self.periodicity
+            grp.attrs["index_order1"] = self.index_order1
+            grp.attrs["index_order2"] = self.index_order2
+
+            for ifile in range(self.nfiles):
+                serial_BAC = self.bitmasks._dumps(ifile)
+                grp.create_dataset(f"nfile_{ifile}", data=np.void(serial_BAC))
```

Comment:
Would you be opposed to making this 0-padded out to like 5 or something?  Maybe that's unnecessary, but it was the first thing I thought of when looking at `h5ls`.

Reply:
fixed in https://github.com/yt-project/yt/pull/4307/commits/89d635a5282c6ddd4f7bac4d580379cf44103405